### PR TITLE
Music: reduce rendering

### PR DESCRIPTION
**Pull request by breville**
This dramatically reduces the number of times the `simple2` timeline elements are rendered.  We now do a single render when starting to play, and another when stopping, which is a huge improvement over rendering upon every tick when playing.

It also makes a few other improvements:
- [limits](https://github.com/code-dot-org/code-dot-org/pull/62637/files#diff-80a97ee2e85a7211b5ed87e04e3d8d0d44ed58fc20ee8de24338b0a26fea1e3aR56) function bounds calculations to a recursion depth of 10.  This is an expensive and unoptimized algorithm, and it makes sense to limit it for a situation that is pretty likely an infinite recursion.
- removes a [sort](https://github.com/code-dot-org/code-dot-org/pull/62637/files#diff-d868892d054aad86c75b4d36be2dfaab2defec4b7491c1b36186b6fab34a19c6L290-R290) which appeared to inadvertently sort the `when run` function bounds below other bounds, causing a function bound rectangle to sometimes be rendered around the `when run` code.  
- handles a [case](https://github.com/code-dot-org/code-dot-org/pull/62637/files#diff-80a97ee2e85a7211b5ed87e04e3d8d0d44ed58fc20ee8de24338b0a26fea1e3aR78-R83) in which functions that call other functions but never emit any sound events could cause a mis-render on the timeline, more likely now that we limit the number of sound events.  
- [fixes](https://github.com/code-dot-org/code-dot-org/pull/62637/files#diff-d868892d054aad86c75b4d36be2dfaab2defec4b7491c1b36186b6fab34a19c6L335-R333) an off-by-1 when limiting events.

### CODE REVIEW

**Comment by mikeharv** [comment]
Should we we do this for Advanced mode too? I'd think there is just as much potential for student code to create very many sounds.

**Review by sanchitmalhotra126 — COMMENTED** [review]
Thanks for digging into this! I took another look at the code because I was a little confused why we were re-rendering every frame in the first place, because we already are pretty good about only recalculating when dependencies changes (e.g. `playbackEvents`), which in theory already should have resulted in the behavior you've implemented here...

I _think_ the culprit actually might be the fact that we pass [this object](https://github.com/code-dot-org/code-dot-org/blob/e4708f90256b2d233c178da4cb88565ccf8404dc/apps/src/music/views/Timeline.tsx#L98) as the props to the `TimelineSimple2Events` view. Since it's just a variable created in the `Timeline` component, I think it may be re-triggering a render on every frame since it gets recreated every render. Additionally, it looks like all the props are just constant values and one function declared outside of the Timeline component. So as an alternate solution, can we try moving/exporting all those values as standalone constants, and then import them directly in `TimelineSimple2Events`, rather than passing them through props? I'm curious if that helps cut down on renders in the same way.

**Comment by breville** [comment]
> I'm curious if that helps cut down on renders in the same way.

It turns out that functional components aren't as pure as the old `PureComponent`, but wrapping the component in a `memo` seems to work pretty well.  We get a few more renders than is ideal, especially when modifying Blockly blocks, but the simplicity is nice.  Most importantly, we now do a single render when starting to play, and another when stopping, which is a huge improvement over rendering every tick when playing.

**Comment by breville** [comment]
> Should we we do this for Advanced mode too? I'd think there is just as much potential for student code to create very many sounds.

Probably yes, but I would like to keep this change scoped given how close we are to limiting deploys before Hour of Code.

File: apps/src/music/views/TimelineSimple2Events.tsx [hunk 1]
```diff
@@ -226,25 +233,35 @@ const TimelineSimple2Events: React.FunctionComponent<
...
+          left={paddingOffset + barWidth * (eventData.when - 1)}
+        />
+      ))}
+    </div>
+  );
+
+  return (
+    <div id="timeline-events">
+      {timelineFunctionExtents}
+      {timelineSoundEvents}
```
Comment (sanchitmalhotra126) [inline apps/src/music/views/TimelineSimple2Events.tsx:262]: Do we still need to break these out if they're no longer memoized?

File: apps/src/music/views/MusicView.jsx [hunk 2]
```diff
@@ -568,6 +568,7 @@ class UnconnectedMusicView extends React.Component {
     this.props.addOrderedFunctions({
       orderedFunctions: this.sequencer.getOrderedFunctions?.() || [],
     });
+
```
Comment (sanchitmalhotra126) [inline apps/src/music/views/MusicView.jsx:571]: nit: remove?

**Review by sanchitmalhotra126 — APPROVED** [review]
Looks good, nice improvement!

**Pull request merged**